src/analysis/monotone-analyzer-impl.h

```diff
@@ -0,0 +1,126 @@
+#ifndef wasm_analysis_monotone_analyzer_impl_h
+#define wasm_analysis_monotone_analyzer_impl_h
+
+#include <iostream>
+#include <unordered_map>
+
+#include "monotone-analyzer.h"
+
+namespace wasm::analysis {
+template<size_t N>
+BlockState<N>::BlockState(const BasicBlock* underlyingBlock)
+  : index(underlyingBlock->getIndex()), cfgBlock(underlyingBlock),
+    states(underlyingBlock->size() + 1, BitsetPowersetLattice<N>::getBottom()),
+    currIndex(0) {}
+
+template<size_t N> void BlockState<N>::addPredecessor(BlockState* pred) {
+  predecessors.push_back(pred);
+}
+
+template<size_t N> void BlockState<N>::addSuccessor(BlockState* succ) {
+  successors.push_back(succ);
+}
+
+template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getFirstState() {
+  return states.front();
+}
+
+template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getLastState() {
+  return states.back();
+}
+
+template<size_t N> void BlockState<N>::visitLocalSet(LocalSet* curr) {
+  states[currIndex].value[curr->index] = true;
+}
+
+template<size_t N> void BlockState<N>::visitLocalGet(LocalGet* curr) {
+  states[currIndex].value[curr->index] = true;
+}
+
+template<size_t N> void BlockState<N>::transfer(std::queue<Index>& worklist) {
+  if (states.size() > 1) {
+    auto cfgIter = cfgBlock->rbegin();
+    currIndex = states.size() - 2;
```

Comment:
Can we do this without using both indices and iterators? It's generally simpler to use just one or the other.

Reply:
I need the iterator to iterate through the CFG block's expressions, while the index is needed to iterate/index the vector of states. The index is a member variable because it lets me access the current state in the visitor without having to change the `Visitor` API.

This is rendered obsolete after we switched to only tracking the beginning and end states for a CFG block instead of tracking state for every expression.